# Проект SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Наша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Задачи

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг — учитывать только книги с 50 и более оценками;
- Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [1]:
# импортируем библиотеки

import pandas as pd
from sqlalchemy import create_engine

# устанавливаем параметры

db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

# сохраняем коннектор

engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
query = '''
        SELECT *
        FROM books
        '''

In [3]:
books = pd.io.sql.read_sql(query, con = engine)
pd.options.display.max_colwidth = 150
books.head()


book_id  author_id                                                  title  \
0        1        546                                           'Salem's Lot   
1        2        465                     1 000 Places to See Before You Die   
2        3        407    13 Little Blue Envelopes (Little Blue Envelope  #1)   
3        4         82  1491: New Revelations of the Americas Before Columbus   
4        5        125                                                   1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [4]:
query2 = '''
        SELECT *
        FROM authors
        '''

In [5]:
authors = pd.io.sql.read_sql(query2, con = engine)
pd.options.display.max_colwidth = 150
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
query3 = '''
        SELECT *
        FROM publishers
        '''

In [7]:
publishers = pd.io.sql.read_sql(query3, con = engine)
pd.options.display.max_colwidth = 150
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
query4 = '''
        SELECT *
        FROM ratings
        '''

In [9]:
ratings = pd.io.sql.read_sql(query4, con = engine)
pd.options.display.max_colwidth = 150
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [10]:
query5 = '''
        SELECT *
        FROM reviews
        '''

In [11]:
reviews = pd.io.sql.read_sql(query5, con = engine)
pd.options.display.max_colwidth = 150
reviews.head(10)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                                                                                               text  
0                                        Mention society tell send professor analysis. Over provide race technology continue these.  
1                                                 Foot glass pretty audience hit themselves. Among admit investment argue security.  
2  Listen treat keep worry. Miss husband tax but person sport treatment industry. Kitchen decision deep the. Social party body the.  
3          Finally month interesting blue could nature cultural bit. Prepare beat finish grow that smile teach. Dream me play near.  
4                  Nation purpose heavy give wait song will. List dinner another whole positive radio fast. Music staff many green.  
5                                                                                          Analysis no several cause international.  
6                              One there cost another. Say type save. With personal where occur direction foot. Half sit role want.  
7                                                                 Within enough mother. There at system full recent play recognize.  
8                                  Thank now focus realize economy focus fly. Item step fine war. Western service history anything.  
9            Game push lot reduce where remember. Including song travel hair foot. Director always myself past story chance reason.

In [12]:
display(reviews.groupby('book_id').agg({'review_id': 'count'}).rename(columns={'review_id': 'count_rev'}).sort_values(by='count_rev', ascending=False))

count_rev
book_id           
948              7
733              6
497              6
696              6
695              6
...            ...
413              1
186              1
855              1
92               1
823              1

[994 rows x 1 columns]

### Задача 1

In [13]:
# Посчитаем, сколько книг вышло после 1 января 2000 года

query_1 = '''
        SELECT book_id, publication_date
        FROM books
        WHERE publication_date > '01-01-2000'
        '''

In [14]:
print('Количество книг:', pd.io.sql.read_sql(query_1, con = engine)['book_id'].count())
print('Крайняя дата публикации:', books['publication_date'].max())

Количество книг: 819
Крайняя дата публикации: 2020-03-31


<div class="alert alert-info">
    
- количество книг, опубликованных после 1 января 2000 г.: 819
- с 2000 года в среднем (по базе данных) публиковалось 40 книг в год
    
<div>

### Задача 2

In [15]:
# Для каждой книги посчитаем количество обзоров и среднюю оценку

query_2 = '''
        with
        books_rating AS (SELECT b.book_id,
                                AVG(ra.rating) AS avg_rating
                        FROM books AS b
                        LEFT JOIN ratings AS ra on b.book_id=ra.book_id
                        GROUP BY b.book_id),
        books_reviews AS (SELECT b.book_id,
                                 COUNT(r.review_id) AS count_reviews
                          FROM books AS b
                          LEFT JOIN reviews AS r on b.book_id=r.book_id
                          GROUP BY b.book_id)
        SELECT title, avg_rating, count_reviews
        FROM books AS b
        LEFT JOIN books_rating ON books_rating.book_id=b.book_id
        LEFT JOIN books_reviews ON books_reviews.book_id=b.book_id
        ORDER BY count_reviews DESC
        '''

In [16]:
book_reviews = pd.io.sql.read_sql(query_2, con = engine)
display(book_reviews.head(10))

print('Количество книг в талице Books:', books['book_id'].nunique())
print('Количество книг в запросе:', book_reviews['title'].count())
print('Дубликаты:', pd.io.sql.read_sql(query_2, con = engine).duplicated().sum())

title  avg_rating  \
0                                     Twilight (Twilight  #1)    3.662500   
1           The Curious Incident of the Dog in the Night-Time    4.081081   
2  Harry Potter and the Chamber of Secrets (Harry Potter  #2)    4.287500   
3                      The Da Vinci Code (Robert Langdon  #2)    3.830508   
4                                            The Glass Castle    4.206897   
5                                             Eat  Pray  Love    3.395833   
6                                      The Catcher in the Rye    3.825581   
7                         The Hobbit  or There and Back Again    4.125000   
8                                   Outlander (Outlander  #1)    4.125000   
9                                                    The Road    3.772727   

   count_reviews  
0              7  
1              6  
2              6  
3              6  
4              6  
5              6  
6              6  
7              6  
8              6  
9              6

Количество книг в талице Books: 1000
Количество книг в запросе: 1000
Дубликаты: 0


<div class="alert alert-info">
    
- наибольшее количество обзоров имеет книга Twilight (Сумерки.Сага.Часть 1)
- почти все книги из топ-10 были экранизированы, и имеют рекордные кассовые сборы в кинотеатрах
- 8 из них имеют несколько частей, являются франшизами, как в книгах, так и в экранизациях
- средний рейтинг топ-10 не ниже 3.5
- топ имеет сотни обзоров благодаря популярности, серийности сюжета и сравнению с экранизацией, в т.ч. "подогреваемые" огромными рекламными бюджетами.
    
<div>

### Задача 3

In [19]:
# Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры

query_3 = '''
        SELECT p.publisher_id,
               p.publisher,
               COUNT(b.book_id) AS count_books
               
        FROM publishers AS p
        LEFT JOIN books AS b on p.publisher_id=b.publisher_id
        WHERE b.num_pages > 50
        GROUP BY p.publisher_id
        ORDER BY count_books DESC
        LIMIT 10;
        '''

In [20]:
display(pd.io.sql.read_sql(query_3, con = engine))

publisher_id                 publisher  count_books
0           212             Penguin Books           42
1           309                   Vintage           31
2           116  Grand Central Publishing           25
3           217          Penguin Classics           24
4            33          Ballantine Books           19
5            35                    Bantam           19
6            45                   Berkley           17
7           284        St. Martin's Press           14
8            46             Berkley Books           14
9            83                     Delta           13

<div class="alert alert-info">
    
- наибольшее количество книг опубликовали Penguin Books - 42
- большая часть топа-10 являются основными издателями Amazon Books
- более 20 книг опубликовали лишь 4 издателя, это в среднем более 1 книги в год (по базе данных)
    
<div>

### Задача 4

In [27]:
# Определим автора с самой высокой средней оценкой книг — учтем только книги с 50 и более оценками

query_4 = '''
        with
        book_authors AS (SELECT a.author,
                                b.book_id
                        FROM authors AS a
                        LEFT JOIN books AS b on b.author_id=a.author_id),
                        
        book_scores AS (SELECT book_id,
                               COUNT(rating_id) AS count_scores
                         FROM ratings
                         GROUP BY book_id
                         HAVING COUNT(rating_id) >= 50
                         )
                
        SELECT author, AVG(r.rating) AS avg_rating
        FROM book_authors
        INNER JOIN book_scores ON book_scores.book_id=book_authors.book_id
        LEFT JOIN ratings AS r ON book_authors.book_id=r.book_id
        GROUP BY author
        ORDER BY avg_rating DESC
        LIMIT 1;
        '''

In [28]:
display(pd.io.sql.read_sql(query_4, con = engine))

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

<div class="alert alert-info">
    
- J.K. Rowling/Mary GrandPré - автор с максимальным средним рейтингом всех своих книг (4.28), имеющих более 50 оценок
    
<div>

### Задача 5

In [23]:
# Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

query_5 = '''
        with
        c_score AS (SELECT username
                    FROM ratings
                    GROUP BY username
                    HAVING COUNT(rating_id) > 50)
        ,
        c_review AS (SELECT username,
                            COUNT(review_id) AS count_reviews
                     FROM reviews
                     GROUP BY username)
        SELECT AVG(c_review.count_reviews) AS avg_reviews
        FROM c_score
        LEFT JOIN c_review ON c_score.username=c_review.username
        ORDER BY avg_reviews DESC
        LIMIT 10;               
        
        '''

In [24]:
display(pd.io.sql.read_sql(query_5, con = engine))

avg_reviews
0    24.333333

<div class="alert alert-info">
    
- среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24
- в среднем эти читатели оставляют 50% обзоров, на общее количество прочитанных книг
    
<div>

### Общие выводы:

###### - количество книг, опубликованных после 1 января 2000 г.: 819
###### - с 2000 года в среднем (по базе данных) публиковалось 40 книг в год

- наибольшее количество обзоров имеет книга Twilight (Сумерки.Сага.Часть 1)
- почти все книги из топ-10 были экранизированы, и имеют рекордные кассовые сборы в кинотеатрах
- 8 из них имеют несколько частей, являются франшизами, как в книгах, так и в экранизациях
- средний рейтинг топ-10 не ниже 3.5
- топ имеет сотни обзоров благодаря популярности, серийности сюжета и сравнению с экранизацией, в т.ч. "подогреваемые" огромными рекламными бюджетами.

###### - наибольшее количество книг опубликовали Penguin Books - 42
###### - большая часть топа-10 являются основными издателями Amazon Books
###### - более 20 книг опубликовали лишь 4 издателя, это в среднем более 1 книги в год (по базе данных)

- J.K. Rowling/Mary GrandPré - автор с максимальным средним рейтингом всех своих книг (4.28), имеющих более 50 оценок

###### - среднее количество обзоров от пользователей, которые поставили больше 50 оценок - 24
###### - в среднем эти читатели оставляют 50% обзоров, на общее количество прочитанных книг.